network: move from operator-specific CRD to cluster config

squeed opened the pull request:
* Add the Network.config.openshift.io CRD
* Generate the network config from the install config
* Change the install config schema to match the cluster config

squeed commented:
/hold
This won't pass CI until https://github.com/openshift/cluster-network-operator/pull/47 merges

squeed commented:
@wking this is the replacement to that previous PR.

abhinavdahiya commented on pkg/types/installconfig.go:
```diff
@@ -106,20 +105,23 @@ func (p *Platform) Name() string {
 
 	// ClusterNetworks is the IP address space from which to assign pod IPs.
-	ClusterNetworks []netopv1.ClusterNetwork `json:"clusterNetworks,omitempty"`
+	ClusterNetwork []ClusterNetworkEntry `json:"clusterNetworks,omitempty"`
```
why is this not list of https://github.com/openshift/api/blob/master/config/v1/types_network.go#L58 ?

squeed replied:
Up to you. The cleanest would probably just be to use `configv1.NetworkSpec` directly. It depends on what you would like the installer to look like. Keep in mind that `NetworkSpec` may slowly change over time (additive only, of course).

abhinavdahiya commented:
> @abhinavdahiya @wking ping on this. I'd like to resolve the discussion of embedding vs. copying the `Network.config.openshift.io` datatype.

I am in favor of keeping them separate and derive Network.config.openshift.io from install-config parts.

also I don't think we should expand clusternetwork to a list, that keeps the installation simple and defers to day-2 for customization.

squeed commented:
@abhinavdahiya we have to support multiple cidrs from day-1 - there are existing use-cases for it, and we don't yet support increasing them.

I can update the PR to just copy the objects.

abhinavdahiya commented:
> @abhinavdahiya we have to support multiple cidrs from day-1 - there are existing use-cases for it, and we don't yet support increasing them.
> 
> I can update the PR to just copy the objects.

+1 on clusternetwork being a list and also a installer type that converts to config.openshift.io type.

abhinavdahiya commented:
with feature freeze and install-config versioned to `v1beta1` https://github.com/openshift/installer/commit/657cb8cd803b71af5616887a7f96ca24b0edb28f this PR needs 
1. exception
2. version bump to v1beta2
2. migration from v1beta1 to v1beta2

/cc @crawford for tracking.

squeed commented:
@abhinavdahiya on it. Should I create a completely separate v1beta1 type for the beta1 -> beta2 migration, or just use optional fields?

squeed commented:
OK, this is updated. It doesn't change the serialized InstallConfig type (though it does change the go representation). Not using cluster.openshift.io is a bug, so this doesn't need an exception (I asked for one, and was told it's not necessary).

@abhinavdahiya ptal, thanks!

squeed commented on pkg/validate/validate.go:
```diff
@@ -21,27 +19,6 @@ var (
 	}()
-
-	// ValidNetworkTypes is a collection of the valid network types.
-	ValidNetworkTypes = map[netopv1.NetworkType]bool{
```
Maybe. If users want to define their own networking (e.g. Calico), then they need to opt-out of the operator. However, there's still some value in them setting the network type to Calico and not None.

squeed replied:
We discussed this for a while when designing the API object and decided that all values are "allowed" for network type, just that the operator won't react to them.

squeed commented on pkg/asset/manifests/network.go:
```diff
@@ -22,13 +22,34 @@ var (
-	// configuration instance of it.
-	// Other operators have their CRD created by the CVO, but we manually
-	// create our operator's configuration in the installer.
+	// We need to manually create our CRDs first, so we can create the
```
I can move the CRDs in to a separate template manifest... up to y'all.

abhinavdahiya replied:
I get an impression from #943 that we do.

squeed commented:
/retest (it's a flake)

knobunc commented:
/retest
(@squeed the bot is finnicky and won't take words after the command)

crawford commented:
> this is making changes to install-config go definition but i think it makes no change to the serialized form... so not sure where this falls on version change...

The serialized form is what we are versioning. If the changes are internal, we don't need to bump the version.

abhinavdahiya commented on data/data/manifests/openshift/cluster-networkconfig-crd.yaml:
```diff
@@ -0,0 +1,19 @@
+# This is the advanced network configuration CRD
```
This should not be created by installer right? only `network.config.openshift.io` 

Looks like we are dropping the `netopv1.NetworkConfig` from vendor as installer no longer creates that...

squeed replied:
I kept it here to make life easier for users who need to create advanced network settings; mind if we keep it for now and see how the docs shake out?

squeed commented:
/retest
test failures are flakes.

abhinavdahiya commented:
/approve

squeed commented:
/retest

abhinavdahiya commented:
Looks like network operator is failing
from https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/1013/pull-ci-openshift-installer-master-e2e-aws/3210/artifacts/e2e-aws/clusteroperators.json
```json
{
    "apiVersion": "v1",
    "items": [
        {
            "apiVersion": "config.openshift.io/v1",
            "kind": "ClusterOperator",
            "metadata": {
                "creationTimestamp": "2019-01-28T17:25:00Z",
                "generation": 1,
                "name": "openshift-network-operator",
                "resourceVersion": "2475",
                "selfLink": "/apis/config.openshift.io/v1/clusteroperators/openshift-network-operator",
                "uid": "a45bfb2b-2321-11e9-ae69-121cbe740854"
            },
            "spec": {},
            "status": {
                "conditions": [
                    {
                        "lastTransitionTime": "2019-01-28T17:46:58Z",
                        "message": "could not apply (networkoperator.openshift.io/v1, Kind=NetworkConfig) /default: could not create (networkoperator.openshift.io/v1, Kind=NetworkConfig) /default: NetworkConfig.networkoperator.openshift.io \"default\" is invalid: []: Invalid value: map[string]interface {}{\"apiVersion\":\"networkoperator.openshift.io/v1\", \"kind\":\"NetworkConfig\", \"metadata\":map[string]interface {}{\"name\":\"default\", \"creationTimestamp\":\"2019-01-28T17:46:58Z\", \"generation\":1, \"uid\":\"b5e79a1e-2324-11e9-a265-121cbe740854\"}, \"spec\":map[string]interface {}{\"additionalNetworks\":interface {}(nil), \"clusterNetworks\":[]interface {}{map[string]interface {}{\"cidr\":\"10.128.0.0/14\", \"hostSubnetLength\":9}}, \"defaultNetwork\":map[string]interface {}{\"type\":\"OpenshiftSDN\", \"otherConfig\":interface {}(nil)}, \"serviceNetwork\":\"172.30.0.0/16\"}, \"status\":map[string]interface {}{}}: validation failure list:\nspec.additionalNetworks in body must be of type array: \"null\"\nspec.defaultNetwork.otherConfig in body must be of type object: \"null\"",
                        "reason": "ApplyClusterConfig",
                        "status": "True",
                        "type": "Failing"
                    },
                    {
                        "lastTransitionTime": "2019-01-28T17:25:01Z",
                        "reason": "Failing",
                        "status": "False",
                        "type": "Progressing"
                    },
                    {
                        "lastTransitionTime": "2019-01-28T17:25:01Z",
                        "reason": "Failing",
                        "status": "False",
                        "type": "Available"
                    }
                ],
                "extension": null,
                "version": ""
            }
        }
    ],
    "kind": "List",
    "metadata": {
        "resourceVersion": "",
        "selfLink": ""
    }
}
```

@squeed ?

squeed commented:
@danwinship check out that validation failure.

danwinship commented:
> validation failure list:
> spec.additionalNetworks in body must be of type array: "null"
> spec.defaultNetwork.otherConfig in body must be of type object: "null"

Doh. Those fields aren't marked `omitempty` in the golang structs, so if you generate your NetworkConfig in golang and then serialize it to JSON, then those fields will be output as `null` rather than being omitted, and then they don't pass the new CRD validation.

abhinavdahiya commented:
> > validation failure list:
> > spec.additionalNetworks in body must be of type array: "null"
> > spec.defaultNetwork.otherConfig in body must be of type object: "null"
> 
> Doh. Those fields aren't marked `omitempty` in the golang structs, so if you generate your NetworkConfig in golang and then serialize it to JSON, then those fields will be output as `null` rather than being omitted, and then they don't pass the new CRD validation.



> > validation failure list:
> > spec.additionalNetworks in body must be of type array: "null"
> > spec.defaultNetwork.otherConfig in body must be of type object: "null"
> 
> Doh. Those fields aren't marked `omitempty` in the golang structs, so if you generate your NetworkConfig in golang and then serialize it to JSON, then those fields will be output as `null` rather than being omitted, and then they don't pass the new CRD validation.

specifically [these](https://github.com/openshift/cluster-network-operator/blob/8053ff5d00ed1db59cd67f5b0994d0813db14fd0/pkg/apis/networkoperator/v1/networkconfig_types.go#L51-L57) right?

squeed commented:
OtherConfig was already fixed. AdditionalNetworks fixed in https://github.com/openshift/cluster-network-operator/pull/81

abhinavdahiya commented:
/retest

squeed commented:
Gah, https://github.com/openshift/cluster-network-operator/pull/83 fixes it.

squeed commented:
/retest

squeed commented:
All green!

abhinavdahiya commented:
@staebler can you PTAL. just want to make sure this is moving in the right direction based on https://github.com/openshift/installer/pull/890

staebler commented on pkg/asset/manifests/network.go:
```diff
@@ -20,35 +22,18 @@ var (
 // Networking generates the cluster-network-*.yml files.
 type Networking struct {
-	config   *netopv1.NetworkConfig
+	config   *configv1.Network
```
This field will be made public in #890. You can make it public now or wait for that PR.

staebler commented on pkg/asset/manifests/network.go:
```diff
@@ -126,10 +95,15 @@ func (no *Networking) Generate(dependencies asset.Parents) error {
 
+	crdContents := ""
+	for _, crdFile := range crds.Files() {
+		crdContents = fmt.Sprintf("%s\n---\n%s", crdContents, crdFile.Data)
```
Why do we want to put these two into the same file rather than keeping them as two separate manifests? If there is a compelling reason to combine them into the same manifest, then why are they separated in the manifest template?

squeed replied:
Because our `yamllint` configuration doesn't allow for multiple documents in the same template file. There's zero reason for it to be separate files otherwise. It is the two CRDs necessary for creating network configuration.

staebler commented on pkg/asset/templates/content/openshift/cluster-network-crds.go:
```diff
@@ -0,0 +1,69 @@
+)
+
+const (
+	netCRDfilename   = "cluster-network-crd.yaml"
```
Either separate this asset into two separate assets or combine the files into a single file.

squeed replied:
`yamllint` blocks single files. Are you okay with me changing the `yamllint` configuration?

staebler commented on pkg/types/installconfig.go:
```diff
@@ -133,12 +132,16 @@ type Networking struct {
+
+	// The size of blocks to allocate from the larger pool.
+	// This is the length in bits - so a 9 here will allocate a /23.
+	HostSubnetLength uint32 `json:"hostSubnetLength"`
```
Can we make this a signed integer, with validation, to conform to kube API conventions?

https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#primitive-types

squeed replied:
done

staebler commented on pkg/types/validation/installconfig.go:
```diff
@@ -70,35 +69,19 @@ func validateNetworking(n *types.Networking, fldPath *field.Path) field.ErrorLis
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("cidr"), cn.CIDR.String(), "cluster network CIDR must not overlap with serviceCIDR"))
+	}
+	if ones, bits := cn.CIDR.Mask.Size(); cn.HostSubnetLength > uint32(bits-ones) {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("hostPrefix"), cn.HostSubnetLength, "cluster network host prefix must not be larger than CIDR"))
```
`hostPrefix` is not the correct json name for the field: it is `hostSubnetLength`.

squeed replied:
oops, fixed.

squeed commented:
Updated and rebased. Tested, works on my machine.

I'd really like to get this merged. I know that the two CRDs isn't the most sensible, but we need to figure out exactly how the installer will handle customizations. That is a deeper conversation, and I would prefer that it not block this PR.

abhinavdahiya commented:
I think https://github.com/openshift/cluster-config-operator will help installer to remove all `*config.openshift.io` CRDs. so for now its fine as-is.

/lgtm

abhinavdahiya commented:
/hold cancel
